### PR TITLE
Evaluate Student Learning: migration to create UserLevelInteractions table 

### PR DESCRIPTION
**Pull request by Erin007**
[AITT-835](https://codedotorg.atlassian.net/browse/AITT-835)
[Proposal ](https://docs.google.com/document/d/1OWlg1XOHtqV5SGCBOTP5fkGDwrzIaCMJQsBMlImsaoc/edit?tab=t.0)

Right now we have an incomplete picture of how students are engaging with levels. Basically, we have the UserLevels table which includes attempts (proxy for # of times the run button is clicked), time_spent and best_result (generally, pass v fail, highly dependent on level type and validation). 

There are a variety of potentially very powerful use cases for more nuanced understanding of what students are doing on levels. More detailed data about student activity opens possibilities to: 

Share meaningful insights with teachers about their students
Provide in-tool interventions for struggling students 
Inform curriculum iterations to adjust difficulty and scaffolding to better meet student needs
Debug issues students report with the platform 
Shape CS Education research about best practices for student learning & engagement 

As an initial step in the Measures of Learning project, this PR sets up a table to store data about how students are interacting with levels. 

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

File: dashboard/db/migrate/20241024180714_create_user_level_interactions.rb [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+      t.integer :level_id, null: false
+      t.integer :script_id, null: false
+      t.string :school_year, null: false
+      t.string :interaction, null: false
+      t.string :code_version
+      t.json :metadata
+
+      t.timestamps
+
+      t.index :user_id
```
Comment (davidsbailey) [inline dashboard/db/migrate/20241024180714_create_user_level_interactions.rb:15]: this index is probably fine for now. once you have a better idea of how you plan to query this table in dashboard (ignoring redshift), you might want to turn this into a composite index like we do here: https://github.com/code-dot-org/code-dot-org/blob/af85735f05528ed4e3ecdd6f8b66a0a840ffeade/dashboard/db/schema.rb#L2253-L2252

File: dashboard/db/migrate/20241024180714_create_user_level_interactions.rb [hunk 2]
```diff
@@ -0,0 +1,18 @@
+class CreateUserLevelInteractions < ActiveRecord::Migration[6.1]
+  def change
+    create_table :user_level_interactions do |t|
+      t.integer :user_id, null: false
+      t.integer :level_id, null: false
+      t.integer :script_id, null: false
+      t.string :school_year, null: false
+      t.string :interaction, null: false
+      t.string :code_version
+      t.json :metadata
```
Comment (davidsbailey) [inline dashboard/db/migrate/20241024180714_create_user_level_interactions.rb:10]: TLDR: I would recommend adding a deleted_at column here.

if we are going to start recording these for real students, then we'll need to be able to delete these as well (see DeleteAccountsHelper) within 30 days of starting to record them. The standard practice is to enable soft-delete via `deleted_at` column and `acts_as_paranoid` directive.

**Review by davidsbailey — COMMENTED** [review]
will there be a model for this table? If so I would recommend creating a model file and rerunning the migration to let rails annotate it as part of this PR.

File: dashboard/db/migrate/20241024180714_create_user_level_interactions.rb (hunk 1 above)
Comment (Erin007) [reply dashboard/db/migrate/20241024180714_create_user_level_interactions.rb:15]: noted!

**Comment by Erin007** [comment]
> will there be a model for this table? If so I would recommend creating a model file and rerunning the migration to let rails annotate it as part of this PR.

Added ✔️

**Review by davidsbailey — APPROVED** [review]

**Pull request merged**